core/modules/query/QueryContext.cc

```diff
@@ -31,19 +31,122 @@
 namespace query {
 
-/// Resolve a column ref to a concrete (db,table)
-/// @return the concrete (db,table), based on current context.
-DbTablePair
-QueryContext::resolve(std::shared_ptr<ColumnRef> cr) {
-    if (!cr) { return DbTablePair(); }
+/// Get the table schema for the tables mentioned in the SQL 'FROM' statement.
+/// This should be adequate and possibly desirable as this information is being used
+/// to restrict queries to particular nodes via the secondary index. Sub-queries are not
+/// supported and even if they were, it could be difficult to determine if a restriction
+/// in a sub-query would be a valid restriction on the entire query.
+void QueryContext::collectTopLevelTableSchema(FromList& fromList) {
+    using ColumnSet=std::set<std::string>;
+    std::map<DbTablePair, ColumnSet> tableColumnMap;
+    for (TableRef::Ptr tblRefPtr : fromList.getTableRefList()) {
+        DbTablePair dbTblPair(tblRefPtr->getDb(), tblRefPtr->getTable()); // DbTablePair has comparisons defined.
+        if (!dbTblPair.db.empty() && !dbTblPair.table.empty()) {
+            // Get the columns in the table from the DB schema and put them in the tableColumnMap.
+            auto columns = getTableSchema(dbTblPair.db, dbTblPair.table);
+            if (!columns.empty()) {
+                ColumnSet& colSet = tableColumnMap[dbTblPair];
+                for (auto const& col : columns) {
+                    colSet.insert(col);
+                }
+            }
+        }
+    }
+
+    // Use the tableToColumnMap to create the columnToTablesMap.
+    columnToTablesMap.clear();
+    for (auto const& tblToColPair : tableColumnMap) {
+        auto const& dbTbl = tblToColPair.first;
+        auto const& colSet = tblToColPair.second;
+        for (auto const& col : colSet) {
+            DbTableSet& st = columnToTablesMap[col];
+            st.insert(dbTbl);
+        }
+    }
+}
+
+
+std::string QueryContext::columnToTablesMapToString() const {
+    std::string str;
+    for (auto const& elem : columnToTablesMap) {
+        str += elem.first + "( "; // column name
+        DbTableSet const& dbTblSet = elem.second;
+        for (auto const& dbTbl : dbTblSet) {
+            str += dbTbl.db + "." + dbTbl.table + " ";
+        }
+        str += ") ";
+    }
+    return str;
+}
+
+///
+/// Get the table schema from the mysqlSchemaConfig database. Primarily, this is
+/// used to map column names to particular tables.
+std::vector<std::string> QueryContext::getTableSchema(std::string const& dbName,
+                                                      std::string const& tableName) {
+    // Check if it is in our list of table schemas
+    DbTablePair dTPair(dbName, tableName);
+
+    // Get the table schema from the local database.
+    std::vector<std::string> colNames;
+    mysql::MySqlConnection mysqlConn{mysqlSchemaConfig};
+    if (!mysqlConn.connected()) {
+        LOGS(_log, LOG_LVL_ERROR, "QueryContext::getTableSchema failed to connect to schema database!");
```

Comment:
But what if this fails outside unit test? I think unit test failures should be handled separately, I'd prefer that in production this failure was not suppressed but resulted in user-visible error.